### PR TITLE
Add be and fe code for derived keys

### PR DESCRIPTION
**Pull request by ArmandoAmoros**
Added functionality to generate a derivedKey given a deviceID.

This key is generated in the back end by hashing together the deviceID and the Enrollment Group Private Key as described [here](https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation?tabs=azure-cli).

Currently the Enrollment Group Private Key is hardcoded as a dummy in the back end for demonstration purposes. In a following step and as part of #54 the key has to be stored safely in the back end server.

MicrosoftTeams-image

### CODE REVIEW

**Comment by ArmandoAmoros** [comment]
@yohny Please have a look to the implementation and feel totally free to rip it up. It is the first time I program in Angular and C#, so all comments would be totally appreciated.

File: hh-dust-fe-be/Frontend/ClientApp/src/app/register/register.component.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+})
+export class RegisterComponent {
+  public registerForm = this.formBuilder.group({
+    deviceId: '',
+  });
+
+  public derivedKey = '';
+
+  public constructor(
+    private formBuilder: FormBuilder,
```
Comment (yohny) [inline hh-dust-fe-be/Frontend/ClientApp/src/app/register/register.component.ts:18]: use underscore as prefix for private fields

File: hh-dust-fe-be/Frontend/Controllers/DeviceRegister.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    public class DeviceRegister : ControllerBase
+    {
+        public DeviceRegister()
+        {
+        }
+
+        [Route("{id}/derivedKey")]
+        public DeviceRegisterDto derivedKey(string id)
+        {
+            byte[] masterKey = Convert.FromBase64String("8isrFI1sGsIlvvFSSFRiMfCNzv21fjbE/+ah/lSh3lF8e2YG1Te7w1KpZhJFFXJrqYKi9yegxkqIChbqOS9Egw==");
```
Comment (yohny) [inline hh-dust-fe-be/Frontend/Controllers/DeviceRegister.cs:20]: `masterKey` seems like sensitive information, so it should not be in code, it should be securely stored server side (in Azure the easiest way is environment variable, for local development use [Secret manger](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-6.0&tabs=windows#secret-manager)) and then read it here using Configuration API. Important - now that it was revealed to the whole world it is compromised and we need a new `masterKey`.

**Review by yohny — CHANGES_REQUESTED** [review]

**Comment by yohny** [comment]
Im a bit confused now where the main development should be happenig...is it here in this monopero? we have separate repo for frontend, that has already setup deployment pipeline to Azure https://github.com/ERNICommunity/hh-dust-fe-be and we could move it in IoT organization if needed. I would prefer to continue development there, as this monorepo ties the independent apps together and thats not good - it makes maintenance and deployment setup harder. the idea ofhaving independent apps that dont share any code in one repo is bad, thats not what monorepos are for.

**Review by dniklaus — COMMENTED** [review]
Please consider to follow the propositions by @yohny regarding underscore prefixes for attributes  and secure master key storage.

**Comment by yohny** [comment]
well thats unfortunate :D this goes against monorepo principle as in our case (independent apps) we gain no advantages from monorepo, But we gel all the disadvantages - unrelated changes/code disturbing you in repo and deployment of apps is harder. Also when you want to work on one app (typical case) you need check out whole codebase that is of no interest to your work

File: hh-dust-fe-be/Frontend/Controllers/DeviceRegister.cs (hunk 2 above)
Comment (ArmandoAmoros) [reply hh-dust-fe-be/Frontend/Controllers/DeviceRegister.cs:20]: As the description of the PR states, this is just a dummy key for testing purposes. Actually is the same key the used as an example in the [Azure documentation](https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation?tabs=azure-cli).

The correct key management is to happen in another PR as per #54, where I'm going to follow your recommendations. Thank you very much for your insights.

**Review by yohny — APPROVED** [review]

**Pull request merged**